### PR TITLE
Add basic KML functionality

### PR DESCRIPTION
**Pull request by patrickelectric**

### CODE REVIEW

**Comment by patrickelectric** [comment]
This will solve #4267

**Comment by patrickelectric** [comment]
selection_536
selection_534
selection_535

**Comment by patrickelectric** [comment]
selection_050

**Comment by DonLakeFlyer** [comment]
Some changes:
* Can you move KML code into MissionManager (not quite right, but better than at root)
* Can you follow the pattern for json save where PlanMasterController only creates/saves the outer object and the code for the real object saving is in the appropriate controller. Then move the KML save code to MissionController instead of PlanMasterController. Then you don't need to expose the load and convert items to public from MissionController.
* Can you Save KML... to the last item in the sync popup. That keeps the upload/download, save/load pattern for this first items which I think makes more sense.

Let me me know if any of this doesn't make sense.

**Comment by DonLakeFlyer** [comment]
There are also build breaks in the code

**Comment by patrickelectric** [comment]
[x]-move KML code into MissionManager (not quite right, but better than at root)
[]-follow the pattern for json save where PlanMasterController only creates/saves the outer object and the code for the real object saving is in the appropriate controller. Then move the KML save code to MissionController instead of PlanMasterController. Then you don't need to expose the load and convert items to public from MissionController.
[]-Save KML... to the last item in the sync popup. That keeps the upload/download, save/load pattern for this first items which I think makes more sense.

**Comment by patrickelectric** [comment]
@DonLakeFlyer Now is better ?

**Comment by DonLakeFlyer** [comment]
Yup, that looks like what I was talking about. Let me take another detailed look through and then should be good to merge.

File: src/MissionManager/MissionController.cc [hunk 1]
```diff
@@ -256,6 +257,49 @@ bool MissionController::_convertToMissionItems(QmlObjectListModel* visualMission
...
+    save(missionJson);
+    _loadItemsFromJson(missionJson, visualItems, error);
+    _convertToMissionItems(visualItems, missionItens, this);
+
+    float altitude = missionJson[_jsonPlannedHomePositionKey].toArray()[2].toDouble();
+
+    QString coord;
+    QStringList coords;
+    for(const auto& item : missionItens) {
+        switch(item->command()) {
```
Comment (DonLakeFlyer) [inline src/MissionManager/MissionController.cc]: This code should be change to use the MissionCommonUIInfo data which has meta data for each mission command:
```
    const MissionCommandUIInfo* uiInfo = qgcApp()->toolbox()->missionCommandTree()->getUIInfo(_controllerVehicle, item->command());
    if (uiInfo && uiInfo->specifiesCoordinate() && !uiInfo->standaloneCoordinate())  {
        // Add KML for a mission item which specifies a coordinate that the vehicle flies through
    }
```

**Comment by DonLakeFlyer** [comment]
Also there is still a build break.

**Comment by patrickelectric** [comment]
oh, problems in debug mode, will solve it today.

**Comment by patrickelectric** [comment]
@DonLakeFlyer Now is it ok ?

File: src/MissionManager/MissionController.cc (hunk 1 above)
Comment (DonLakeFlyer) [inline src/MissionManager/MissionController.cc]: This still needs to be fixed.

Comment (patrickelectric) [inline src/MissionManager/MissionController.cc]: Okey dokey, sorry for missed that.

Comment (patrickelectric) [reply src/MissionManager/MissionController.cc]: @DonLakeFlyer Done.

**Comment by patrickelectric** [comment]
@DonLakeFlyer we can merge that ?

**Comment by DonLakeFlyer** [comment]
I tried to save a KML file, but when I loaded back into Google Earth it did zoom into the right location but no lines showed up. Is there something else I need to do in Google Earth.

**Comment by DonLakeFlyer** [comment]
Nevermind, I figure out how to get lines to show. I think there is an altitude calculation problems somewhere, still looking.

**Comment by DonLakeFlyer** [comment]
[KMLTest.plan.zip](https://github.com/mavlink/qgroundcontrol/files/1285546/KMLTest.plan.zip)

Try loading this plan file without a vehicle connected and save to KML. I think the home position altitude is being calculated incorrectly, or maybe the takeoff item. The first point is way up in the air.

**Comment by patrickelectric** [comment]
@DonLakeFlyer no problem here to open it.
image


Are you doing the same thing that this gif ?
https://media.giphy.com/media/l1J9A3v2CMmmFQPza/giphy.gif

**Comment by patrickelectric** [comment]
Now, it appears that have a problem with Home position, ill take a look.

**Comment by patrickelectric** [comment]
image

**Comment by patrickelectric** [comment]
Can you check now @DonLakeFlyer ?

**Comment by DonLakeFlyer** [comment]
Yup, all good now thanks.

**Pull request merged**